tls/tls_srv.c

```diff
@@ -0,0 +1,2331 @@
+		*(unsigned short *)tls->hs->tmp = 0;
+		io->hslen -= 2;
+		/* 1 for comp. alg. len */
+		if (n < 2 || n + 1 > io->hslen || (n & 1)) {
+			T_DBG("ClientHello: bad cipher suite length %d\n", n);
+			ttls_send_alert(tls, TTLS_ALERT_LEVEL_FATAL,
+					TTLS_ALERT_MSG_DECODE_ERROR);
+			return TTLS_ERR_BAD_HS_CLIENT_HELLO;
+		}
+		TTLS_HS_FSM_MOVE(TTLS_CH_HS_CS);
+	}
+
+	/* Read a cipher suite. */
+	T_FSM_STATE(TTLS_CH_HS_CS) {
+		unsigned short cs, *csn = (unsigned short *)tls->hs->tmp;
+		const unsigned int off = TTLS_HS_TMP_STORE_SZ + 2;
+		/* Get number of ciphersuite bytes from the client. */
+		n = ntohs(*(short *)&tls->hs->tmp[TTLS_HS_TMP_STORE_SZ]);
+		BUG_ON(io->rlen >= 2);
+		/*
+		 * Skip the last low-priority cipher suites which we can not
+		 * store.
+		 */
+		if (2 + *csn + 2 > TTLS_HS_CS_MAX_SZ) {
+			/* Write number of stored cipher suite bytes. */
+			*(short *)&tls->hs->tmp[TTLS_HS_TMP_STORE_SZ] = htons(*csn);
+			io->hslen -= n - *csn;
+			TTLS_HS_FSM_MOVE(TTLS_CH_HS_COMPN);
+		}
+		/* Read current cipher suite, just after the counter. */
+		if (unlikely(io->rlen)) {
+			tls->hs->tmp[off + *csn + 1] = *p;
+			p++;
+		}
+		else if (unlikely(buf + len - p == 1)) {
+			tls->hs->tmp[off + *csn] = *p;
+			io->rlen = 1;
+			p++;
+			T_FSM_EXIT();
+		}
+		else {
+			memcpy_fast(&tls->hs->tmp[off + *csn], p, 2);
+			p += 2;
+		}
+		cs = ntohs(*(short *)&tls->hs->tmp[off + *csn]);
+		T_DBG3("ClientHello: cipher suite #%u: %#x\n",
+		       *csn / 2, cs);
+		if (ttls_check_scsv_fallback(tls, cs))
+			return TTLS_ERR_BAD_HS_CLIENT_HELLO;
+		io->hslen -= 2;
+		*csn += 2;
+		if (*csn == n)
+			TTLS_HS_FSM_MOVE(TTLS_CH_HS_COMPN);
+		TTLS_HS_FSM_MOVE(TTLS_CH_HS_CS);
+	}
+
+	/* Check the compression algorithms length. */
+	T_FSM_STATE(TTLS_CH_HS_COMPN) {
+		n = *p;
+		if (n < 1 || n > 16 || n + 1 > io->hslen) {
+			T_DBG("ClientHello: bad compression number %d\n", n);
+			ttls_send_alert(tls, TTLS_ALERT_LEVEL_FATAL,
+					TTLS_ALERT_MSG_DECODE_ERROR);
+			return TTLS_ERR_BAD_HS_CLIENT_HELLO;
+		}
+		/*
+		 * Use 2 bytes to read out compression stuff: 1st for
+		 * compression algorithms to read and 2nd for a flag
+		 * that we've faced NULL-compression.
+		 */
+		tls->hs->tmp[0] = n;
+		tls->hs->tmp[1] = 0;
+		T_DBG3("ClientHello: compression algorithms length %u\n", n);
+		io->hslen--;
+		++p;
+		TTLS_HS_FSM_MOVE(TTLS_CH_HS_COMP);
+	}
+
+	T_FSM_STATE(TTLS_CH_HS_COMP) {
+		if (*p == TTLS_COMPRESS_NULL) {
+			T_DBG3("saw NULL compression\n");
+			tls->hs->tmp[1] = 1;
+		}
+		io->hslen--;
+		++p;
+		if (!--tls->hs->tmp[0]) {
+			if (!tls->hs->tmp[1]) {
+				T_DBG("ClientHello: no NULL compression\n");
+				ttls_send_alert(tls, TTLS_ALERT_LEVEL_FATAL,
+						TTLS_ALERT_MSG_DECODE_ERROR);
+				return TTLS_ERR_BAD_HS_CLIENT_HELLO;
+			}
+			TTLS_HS_FSM_MOVE(TTLS_CH_HS_EXTLEN);
+		}
+		TTLS_HS_FSM_MOVE(TTLS_CH_HS_COMP);
+	}
+
+	/* Read extensins length. */
+	T_FSM_STATE(TTLS_CH_HS_EXTLEN) {
+		/* Use the first 2 bytes for toal number of extensions. */
```

Comment:
Maybe `total` instead of `toal` should be.
In previous comment: `extensins` -> `extensions`.